Objects/listobject.c

```diff
@@ -3174,6 +3174,25 @@ PyList_AsTuple(PyObject *v)
     return ret;
 }
 
+PyObject *
+_PyList_AsTupleAndClear(PyListObject *self)
+{
+    assert(self != NULL);
+    PyObject *ret;
+    if (self->ob_item == NULL) {
+        return PyTuple_New(0);
+    }
+    Py_BEGIN_CRITICAL_SECTION(self);
+    PyObject **items = self->ob_item;
+    Py_ssize_t size = Py_SIZE(self);
+    self->ob_item = NULL;
+    Py_SET_SIZE(self, 0);
+    ret = _PyTuple_FromArraySteal(items, size);
+    free_list_items(items, false);
```

Comment:
If `_PyTuple_FromArraySteal`, do we still want to free the list items?

Reply:
Yes. `_PyTuple_FromArraySteal` steals the references, but doesn't do anything with the memory

Reply:
Yes. `_PyTuple_FromArraySteal` consumes the all references regardless of whether it succeeds or fails.
It would be a pain to use otherwise.